### PR TITLE
Validate types during runtime

### LINKED ISSUE #20: Make sure static arguments are not tracers
Something like this gives a nasty, cryptic traceback:

```python
>>> jax.jit(mpi4jax.Send)(jnp.ones(10), 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/mpi4jax/collective_ops/send.py", line 86, in mpi_send_xla_encode
    _constant_s32_scalar(c, dest),
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/mpi4jax/utils.py", line 68, in _constant_s32_scalar
    return _ops.Constant(c, _np.int32(x))
jax.traceback_util.FilteredStackTrace: Exception: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray(int32[], weak_type=True):JaxprTrace(level=-1/1)>.

This error can occur when a JAX Tracer object is passed to a raw numpy function, or a method on a numpy.ndarray object. You might want to check that you are using `jnp` together with `import jax.numpy as jnp` rather than using `np` via `import numpy as np`. If this error arises on a line that involves array indexing, like `x[idx]`, it may be that the array being indexed `x` is a raw numpy.ndarray while the indices `idx` are a JAX Tracer instance; in that case, you can instead write `jax.device_put(x)[idx]`.

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/jax/traceback_util.py", line 136, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/jax/api.py", line 214, in f_jitted
    donated_invars=donated_invars)
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/jax/core.py", line 1146, in bind
    return call_bind(self, fun, *args, **params)
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/jax/core.py", line 1135, in call_bind
    outs = primitive.impl(fun, *args, **params)
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/jax/interpreters/xla.py", line 530, in _xla_call_impl
    *unsafe_map(arg_spec, args))
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/jax/linear_util.py", line 234, in memoized_fun
    ans = call(fun, *args)
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/jax/interpreters/xla.py", line 648, in _xla_callable
    extend_name_stack(wrap_name(name, 'jit')), *xla_args)
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/jax/interpreters/xla.py", line 408, in jaxpr_subcomp
    ans = rule(c, *in_nodes, **eqn.params)
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/mpi4jax/collective_ops/send.py", line 86, in mpi_send_xla_encode
    _constant_s32_scalar(c, dest),
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/mpi4jax/utils.py", line 68, in _constant_s32_scalar
    return _ops.Constant(c, _np.int32(x))
  File "/Users/dion/.virtualenvs/tempenv-7fe4208272c9d/lib/python3.7/site-packages/jax/core.py", line 455, in __array__
    raise Exception(msg)
Exception: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray(int32[], weak_type=True):JaxprTrace(level=-1/1)>.

This error can occur when a JAX Tracer object is passed to a raw numpy function, or a method on a numpy.ndarray object. You might want to check that you are using `jnp` together with `import jax.numpy as jnp` rather than using `np` via `import numpy as np`. If this error arises on a line that involves array indexing, like `x[idx]`, it may be that the array being indexed `x` is a raw numpy.ndarray while the indices `idx` are a JAX Tracer instance; in that case, you can instead write `jax.device_put(x)[idx]`.
```

We should check in our `xla_encode` implementations if the user passed a value that we expect as static (such as `dest` or `tag`) as a tracer and throw a more useful error.

### PR DESCRIPTION
**Pull request by dionhaefner**
Fixes #20.

With this, the example from #20 gives:

```python
>>> jax.jit(mpi4jax.Send)(jnp.ones(10), 1)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/dion/codes/mpi4jax/mpi4jax/validation.py", line 70, in wrapped
    f'{func_name} got unexpected type for argument "{arg}" '
jax.traceback_util.FilteredStackTrace: TypeError: Send got unexpected type for argument "dest" (expected: int, got: <class 'jax.interpreters.partial_eval.JaxprTracer'>).

An abstract tracer was passed where a concrete value is expected. Try using the static_argnums argument of jax.jit.

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/traceback_util.py", line 136, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/api.py", line 214, in f_jitted
    donated_invars=donated_invars)
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/core.py", line 1146, in bind
    return call_bind(self, fun, *args, **params)
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/core.py", line 1135, in call_bind
    outs = primitive.impl(fun, *args, **params)
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/interpreters/xla.py", line 530, in _xla_call_impl
    *unsafe_map(arg_spec, args))
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/linear_util.py", line 224, in memoized_fun
    ans = call(fun, *args)
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/interpreters/xla.py", line 601, in _xla_callable
    fun, pvals, instantiate=False, stage_out=True, bottom=True)
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/interpreters/partial_eval.py", line 422, in trace_to_jaxpr
    jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
  File "/Users/dion/.virtualenvs/science/lib/python3.7/site-packages/jax/linear_util.py", line 150, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/Users/dion/codes/mpi4jax/mpi4jax/validation.py", line 70, in wrapped
    f'{func_name} got unexpected type for argument "{arg}" '
TypeError: Send got unexpected type for argument "dest" (expected: int, got: <class 'jax.interpreters.partial_eval.JaxprTracer'>).

An abstract tracer was passed where a concrete value is expected. Try using the static_argnums argument of jax.jit.
```

This works by introducing a decorator `enforce_types` that works like this:

```python
>>> @enforce_types(
...     foo=(int, type(None)),
...     bar=str
... )
... def func(x, foo, bar):
...     print(bar)

>>> func(1, 2, 'hi there!')
hi there!

>>> func(1, 2, 3)
TypeError: func got unexpected type for argument "bar" (expected: str, got: <class 'int'>)
```

Since we are passing these objects as raw pointers to C it felt appropriate to be a bit stricter with types than usual.

This also means that something like this throws an error:

```python
>>> func(1, np.uint64(2), 'hi there!')
TypeError: func got unexpected type for argument "foo" (expected: ['int', 'NoneType'], got: <class 'numpy.uint64'>).
```

which is acceptable collateral damage IMO.

### CODE REVIEW

**Comment by dionhaefner** [comment]
Still to-do:

- [ ] Unit tests

**Comment by PhilipVinc** [comment]
> which is acceptable collateral damage IMO.

Are we sure we can't find a workaround to this? Maybe we could manually replace occurrences of int with `int, np.uint32, np.uint64` and so on.
I think it's a feature of jax the possibility of passing arbitrary numpy stuff and get it converted to jax types.

However, you never check the 'array' argument in your pr, so... it seems like you avoid the problem.

File: mpi4jax/collective_ops/sendrecv.py [hunk 1]
```diff
@@ -18,16 +18,27 @@
...
 
 
 # This function applies the primitive to an AST
+@enforce_types(
+    source=int,
+    dest=int,
+    sendtag=int,
+    recvtag=int,
+    comm=(_MPI.Intracomm, HashableMPIType),
+    status=(type(None), _MPI.Status),
```
Comment (PhilipVinc) [inline mpi4jax/collective_ops/sendrecv.py:39]: `status` is never passed as an hashable type?

**Comment by dionhaefner** [comment]
> However, you never check the 'array' argument in your pr, so... it seems like you avoid the problem.

This is just about the non-tracer arguments such as `source` or `tag`. The arrays that are communicated can be any type.

File: mpi4jax/collective_ops/sendrecv.py (hunk 1 above)
Comment (dionhaefner) [reply mpi4jax/collective_ops/sendrecv.py:39]: Good catch! This has to be hashable, too.

**Comment by dionhaefner** [comment]
This now allows NumPy generic types:

```python
@enforce_types(
    x=np.integer
)
```

would allow Python `int` and all NumPy integer types (`np.uint64`,  `np.int32`, ...).

**Review by PhilipVinc — APPROVED** [review]

**Pull request merged**